backend/src/main/java/com/google/step/servlets/FormsServlet.java

```diff
@@ -0,0 +1,250 @@
+            return;
+        }
+
+        //read the content body
+        long formId;
+        String formName;
+        if (request.getContentType().contains("application/x-www-form-urlencoded")) {
+            formId = Long.parseLong(request.getParameter("form_id"));
+            formName = request.getParameter("form_name");
+        } else if (request.getContentType().contains("application/json")) {
+            Gson gson = new Gson();
+            Map<String, String> jsonMap = gson.fromJson(request.getReader(), Map.class);
+            formId = Long.parseLong(jsonMap.get("form_id"));
+            formName = jsonMap.get("form_name");
+        } else {
+            response.sendError(415,
+                    "Content type not supported. Try application/x-www-form-urlencoded or application/json.");
+            return;
+        }
+
+        //query the datastore to see if the form id already is claimed
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("Form")
+                .setFilter(new Query.FilterPredicate("formId", Query.FilterOperator.EQUAL, formId))
+                .setFilter(new Query.FilterPredicate("verified", Query.FilterOperator.EQUAL, true))
+                .setKeysOnly();
+        PreparedQuery queryResults = datastore.prepare(query);
+        if (queryResults.asList(FetchOptions.Builder.withDefaults()).size() > 0) { //there already exists a verified form with this id
+            response.sendError(403, "Form ID already claimed by another user.");
+        }
+
+        String webhookUrl = generateUserWebhook(request, UserAuthenticationUtil.getCurrentUser());
+        String googleKey = generateRandomGoogleKey(20);
+
+        //store the form entity in the datastore
+        Form newForm = new Form(formId,
+                formName,
+                AdvertiserUtil.createAdvertiserKey(UserAuthenticationUtil.getCurrentUser()),
+                googleKey);
+        datastore.put(newForm.asEntity());
+
+        response.setContentType("application/json;");
+        response.getWriter().println(new WebhookResponse(webhookUrl, googleKey, formId).toJson());
+    }
+
+    /**
+     * @return the webhook for this user with a URL-Safe Key String uniquely identifying the user
+     */
+    private String generateUserWebhook(HttpServletRequest request, User user) {
+        //generate URL-Safe Key string
+        String advertiserKeyString = KeyFactory.keyToString(AdvertiserUtil.createAdvertiserKey(user));
+        return request.getScheme() + "://" +
+                request.getServerName() + ":" +
+                request.getServerPort() + "/api/webhook?" + ID_URL_PARAM + "=" +
+                advertiserKeyString;
+    }
+
+    /**
+     * Generates a random Google Key of the specified length
+     * @param length the length of the Google Key
+     * @return       the randomly generated Google Key
+     */
+    private String generateRandomGoogleKey(int length) {
```

Comment:
Yes, I believe I can use Java's UUID class to generate a random google string. Internally it also uses a SecureRandom to generate random strings.
I could accomplish the same thing with UUID.randomUUID().toString().replace("-","").substring(0,20);
However, I believe this UUID generated string would only include hex characters (0-9 and A-F), and it would always be generating ~128 random bits while we wouldn't be using that much.
Should I change this to use the UUID implementation instead?

Alternatively, I could import a non-Java library like Apache RandomUtils however I'm not sure if that's worth just for this one function.

Reply:
that's fine let's keep this for now.

Reply:
okay sounds good